fix: NetworkBehaviour Property Initialization Updates

NoelStephensUnity opened the pull request:
Accessing specific properties of NetworkBehaviour can too easily cause exceptions to occur.  Depending upon frequency of usage, they can also impact performance unnecessarily. 

[MTT-2785](https://jira.unity3d.com/browse/MTT-2785)

## Changelog
### com.unity.netcode.gameobjects
- Fixed: If you update multiple packages, create a new section with a new header for the other package. 

## Testing and Documentation
* One test was updated.
* The rest validate the change

NoelStephensUnity committed:
fix

Updating to more stable and performant NetworkBehaviour properties through changing the how and when the properties are set.

NoelStephensUnity committed:
test fix

Really MessageOrdering and the SpawnRpcDespawn classes need to be reviewed and updated.  This fixes a few timing related issues that were exposed with this update.

NoelStephensUnity committed:
fix

Removing the update to HasNetworkObject

NoelStephensUnity committed:
update

really only ever need to wait for almost 1 tic interval to make sure the GameObject and associated components are destroyed and the handler destroyed.

NoelStephensUnity committed:
style

clarity in a comment

NoelStephensUnity committed:
update

removing the NetworkManager.IsListening since that is part of how the property values are determined (host, client, server).

NoelStephensUnity committed:
update

handling an unlikely condition where the ownership or property update is invoked and NetworkObject is null.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2785-networkbehaviour-property-initialization

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -237,40 +237,41 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
         /// Gets if the object is the the personal clients player object
         /// </summary>
-        public bool IsLocalPlayer => NetworkObject.IsLocalPlayer;
+        public bool IsLocalPlayer => m_NetworkObject != null ? m_NetworkObject.IsLocalPlayer : false;
```
Is this actually an improvement though?  If the NO is null, I think `IsLocalPlayer` should be undefined / access should crash as we actually don't really know.  If there's code later that wants to tolerate not having the NO set up yet, fine, then let that code do this check.  Also, since we call this frequently...

NoelStephensUnity replied:
I went ahead and made this a private set and just initialize it during the spawning/despawning stages.
It removes the processing overhead and is cleaner.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -352,7 +353,32 @@ protected NetworkBehaviour GetNetworkBehaviour(ushort behaviourId)
-        public ulong OwnerClientId => NetworkObject.OwnerClientId;
+        public ulong OwnerClientId { get; internal set; }
+
+        internal void UpdatePropertyStates()
```
```suggestion
        internal void UpdateNetworkProperties()
```

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -366,21 +392,56 @@ protected NetworkBehaviour GetNetworkBehaviour(ushort behaviourId)
         }
 
-        internal void InternalOnNetworkDespawn() { }
+        internal void UpdateOwnership()
+        {
+            if (NetworkObject != null)
+            {
+                IsOwnedByServer = NetworkObject.IsOwnedByServer;
+                IsOwner = NetworkObject.IsOwner;
+                OwnerClientId = NetworkObject.OwnerClientId;
+            }
+            else // Shouldn't happen, but if it does then just clear out everything
+            {
+                IsOwnedByServer = IsOwner = false;
+                OwnerClientId = 0;
+            }
+        }
```
move this logic into `UpdateNetworkProperties()` and maintain updating all net props in one place. easier to maintain.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -366,21 +392,56 @@ protected NetworkBehaviour GetNetworkBehaviour(ushort behaviourId)
         /// </summary>
         public virtual void OnGainedOwnership() { }
 
+        internal void GainedOwnership()
```
```suggestion
        internal void InternalOnGainedOwnership()
```

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -366,21 +392,56 @@ protected NetworkBehaviour GetNetworkBehaviour(ushort behaviourId)
         /// </summary>
         public virtual void OnLostOwnership() { }
 
+        internal void LostOwnership()
```
```suggestion
        internal void InternalOnLostOwnership()
```

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -237,40 +237,41 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
+        public bool IsOwner { get; internal set; }
 
+        private bool m_IsServer;
         /// <summary>
         /// Gets if we are executing as server
         /// </summary>
-        protected bool IsServer => IsRunning && NetworkManager.IsServer;
+        protected bool IsServer { get { return m_IsServer; } }
```
I'd suggest a simpler syntax like:
```
protected bool IsServer { get; private set; }
```
and for other properties too :)

0xFA11 replied:
and you don't need to have another field to back it up.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -366,21 +392,56 @@ protected NetworkBehaviour GetNetworkBehaviour(ushort behaviourId)
 
         internal void InternalOnNetworkSpawn()
```
I like the fact that now we are relying on `OnNetworkSpawn` and `OnNetworkDespawn` now <3

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2785-networkbehaviour-property-initialization

NoelStephensUnity committed:
update

Includes all of Fatih's suggested adjustments
Also, went ahead and just moved the invocation of OnNetworkSpawn and OnNetworkDespawn into the internal methods.

NoelStephensUnity committed:
update

Coming to alternate solution of IsLocalPlayer to reduce the processing cost of the other implementation.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2785-networkbehaviour-property-initialization

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -352,7 +352,47 @@ protected NetworkBehaviour GetNetworkBehaviour(ushort behaviourId)
+            }
+            else // Shouldn't happen, but if so then set the properties to their default value;
+            {
+                OwnerClientId = NetworkObjectId = default;
```
Is this an error condition we should log?  Would be a way to spot mis-use / regressions

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -549,15 +549,15 @@ internal void InvokeBehaviourOnLostOwnership()
             for (int i = 0; i < ChildNetworkBehaviours.Count; i++)
             {
-                ChildNetworkBehaviours[i].OnGainedOwnership();
+                ChildNetworkBehaviours[i].InternalOnGainedOwnership();
```
Did you consider putting the call to OnGainedOwnership here vs inside InternalOnGainedOwnership?

NoelStephensUnity replied:
Actually, the internal accessor method is what should invoke the OnGainedOwnership but not until the "internal stuff" is done first.  It makes the entire ownership flow for NetworkBehaviours occur within NetworkBehaviours.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -352,7 +352,47 @@ protected NetworkBehaviour GetNetworkBehaviour(ushort behaviourId)
+        /// dependencies such as a NetworkObject's spawned
+        /// state or NetworkManager's session state.
+        /// </summary>
+        internal void UpdateNetworkProperties()
```
Did you consider splitting this into multiple functions?  For instance, I don't think you need to re-evaluate NetworkBehaviourId when you change ownership or or despawn.  Then those functions would be more clear about what state updates are tied to which events.  Also, while probably not a major perf thing, you can avoid code that never needs to run again (e.g. IsHost/IsClient/IsServer updates)

NoelStephensUnity replied:
🤣 
Actually I had broken this up originally (see Fatih's suggested changes above) into two methods where the UpdateNetworkProperties was only invoked during spawn and despawn but OnOwnershipUpdate was invoked when ownership changed and during spawn and despawn.

Really, the total processing time incurred by setting everything during spawn, despawn, and ownership change is negligible relative to the way things were originally.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2785-networkbehaviour-property-initialization
